src/NServiceBus.Transport.RabbitMQ/Receiving/MessagePump.cs

```diff
@@ -117,7 +113,7 @@ public async Task Stop()
             consumer.Received -= Consumer_Received;
             messageProcessing.Cancel();
 
-            while (semaphore.CurrentCount != maxConcurrency)
+            while (Volatile.Read(ref numberOfExecutingReceives) != maxConcurrency)
```

Comment:
Is `Volatile.Read` actually required here? I always have to go read up on volatile stuff to recall when it might be needed.

Reply:
Also, I don't think the logic for the condition is correct now. With the semaphore, `CurrentCount == maxConcurrency` means there are no messages being processed.

With the counter approach, we need to wait until it is 0, right?

Reply:
Yeah sorry I wasn't careful enough here and rushed things in between other stuff. Will fix. Sorry for that lousy change

Reply:
Have a look, what condition do you prefer? I currently have `> 0` because I felt it is a more safe value than `!= 0` but I might be overthinking it